### DISCUSSION TITLE
Write once, run everywhere?

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#149] [General]
I don't know if Objeck has ever claimed something like that. But since it's very similar to Java, let's assumed it did. On Java, even though it bumped the class version on each major releases, the newer Java will always able to run the class compiled by older Java versions. It seems to be not the case on Objeck, isn't it? @objeck @rhollines 

This is when I tried to run the precompiled Objeck LSP:

`This executable appears to be invalid or compiled with a different version of the toolchain.`

Just released 2 days ago. Of course, it was compiled by the same Objeck version as what I have (v2023.2.1). So why it's this way?

p/s: Java class versions list: https://www.baeldung.com/java-find-class-version

**Comment by ghost** [comment]
Sorry @objeck @rhollines, I have many Objeck versions on my environment variables (including the ones I compiled locally). I used the wrong Objeck version (older) :face_with_spiral_eyes:

**Comment by objeck** [comment]
It is a safeguard. When new instructions are added (rare nowadays) or TRAP directives (a more likely scenario, but rare), the compiler and VM need to be synced. This could be relaxed in further releases.

**Comment by ghost** [comment]
I compiled my version from source (this means my Objeck is even newer than the release), but it still said `This executable appears to be invalid or compiled with a different version of the toolchain.` This shouldn't be the case. The newer versions should always able to run code compiled by the older versions!

**Reply by ghost** [reply]
https://github.com/objeck/objeck-lang/blob/master/core/vm/loader.cpp#L65
